Let C[w, m] be the cost of transporting one truck-load of EVs from w to m and $x_{w \ m} $ be  number of truck-loads of EVs transported from w to m

where w is warehouses location and m is market location

general optimisation problem is 

min $ \sum_{w=1}^{w=n} \sum_{m=1}^{m=k} C[w,m]*x_{w \ m} $

$ \sum_{w=1}^{w=n} x_{w \ j} =b_j \ \ \  ∀j \in m=(1,2,3,4,.........k) $

$ \sum_{m=1}^{m=k} x_{i \ m} \le c_i \ \ \ ∀i \in w=(1,2,3,4,5,6,........n)$

w=(1,2,3,4,5,6,..........n)

m=(1,2,3,4,5,6,..........k)

$b_j$ is the  demand at $j^{th}$ market

$c_i$ is be available stock at $i^{th}$ location

$x_{w \ m } \ge 0 \ \ \ and \ \ must \ \  be \ \ integer$

i.e $x_{w \ m} \in (0,1,2,3,4,5,..............)$


in our problem

w = ( 1,2,3,4,5,6,7,8) correspoding to warehouse location (Ahmedabad,Bengaluru,Chennai,Delhi,Kolkata,Lucknow,Mumbai,Hyderabad) 

and m= (1,2,3,4,5,6,7,8) corresponding to market location ( Hassan,Jodhpur,Trichy,Simla,Vijayawada,Dibrugarh,Raipur,Agartala)

now our optimisation problem becomes 

min $ \sum_{w=1}^{w=8} \sum_{m=1}^{m=8} C[w,m]*x_{w \ m} $

subjected to 

$ \sum_{w=1}^{w=8} x_{w \ 1} =205$

$ \sum_{w=1}^{w=8} x_{w \ 2} =300$

$ \sum_{w=1}^{w=8} x_{w \ 3} =425$

$ \sum_{w=1}^{w=8} x_{w \ 4} =270$

$ \sum_{w=1}^{w=8} x_{w \ 5} =375$

$ \sum_{w=1}^{w=8} x_{w \ 6} =225$

$ \sum_{w=1}^{w=8} x_{w \ 7} =185$

$ \sum_{w=1}^{w=8} x_{w \ 8} =110$

$ \sum_{m=1}^{m=8} x_{1 \ m} \le350$

$ \sum_{m=1}^{m=8} x_{2 \ m} \le185$

$ \sum_{m=1}^{m=8} x_{3 \ m} \le305$

$ \sum_{m=1}^{m=8} x_{4 \ m} \le180$

$ \sum_{m=1}^{m=8} x_{5 \ m} \le205$

$ \sum_{m=1}^{m=8} x_{6 \ m} \le175$

$ \sum_{m=1}^{m=8} x_{7 \ m} \le325$

$ \sum_{m=1}^{m=8} x_{8 \ m} \le370$

$x_{w \ m } \ge 0 \ \ \ and \ \ must \ \  be \ \ integer$

i.e $x_{w \ m} \in (0,1,2,3,4,5,..............)$






In [ ]:
!pip install -q pyomo

In [ ]:
from pyomo.environ import *

In [ ]:
import numpy as np

In [ ]:
import pandas as pd


In [ ]:
data_csvfile = pd.read_csv('lab6_ex1.csv')

In [ ]:
print(data_csvfile)

   Warehouses  Hassan  Jodhpur  Trichy  Simla  Vijayawada  Dibrugarh  Raipur  \
0  Ahmedabad      827      817    1470   2272        1615       3143    1264   
1   Bengaluru     559     1832     751   3772         637       3773    1856   
2     Chennai    1109     2443     577   3865         798       3521    1984   
3       Delhi    2943     1471    4676   1337        2702       2741    2482   
4     Kolkata    3742     3089    2931   3199        2226       1739    1691   
5     Lucknow    2353     1160    3295    368        2369       2365    1293   
6      Mumbai    1356     1617    2045   2941        2334       4151    1937   
7   Hyderabad     195     1862    1331   2921         402       2825     563   
8      Demand     205      300     425    270         375        225     185   

   Agartala  Available_Stock  
0      4465              350  
1      5459              185  
2      5303              305  
3      4148              180  
4      2128              205  
5      3642  

In [ ]:
model_lab6_ex1= ConcreteModel()

In [ ]:
M=len(data_csvfile.index)-1

In [ ]:
N=len(data_csvfile.columns)-2

In [ ]:
row_indices=range(M)
col_indices = range(N)

In [ ]:
model_lab6_ex1.x = Var(row_indices,col_indices,domain=NonNegativeIntegers)

In [ ]:
model_lab6_ex1.objective=Objective(expr=sum(data_csvfile.iloc[i,j+1]*model_lab6_ex1.x[i,j] for i in row_indices for j in col_indices),sense=minimize)

In [ ]:
model_lab6_ex1.constraints=ConstraintList()

In [ ]:
for i in row_indices:
  model_lab6_ex1.constraints.add(expr=sum(model_lab6_ex1.x[i,j] for j in col_indices) <= data_csvfile.iloc[i,-1])

In [ ]:
for j in col_indices:
  model_lab6_ex1.constraints.add(expr=sum(model_lab6_ex1.x[i,j] for i in row_indices)==data_csvfile.iloc[M,j+1])

In [ ]:
model_lab6_ex1.pprint()

4 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   16 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16}
    x_index : Size=1, Index=None, Ordered=True
        Key  : Dimen : Domain              : Size : Members
        None :     2 : x_index_0*x_index_1 :   64 : {(0, 0), (0, 1), (0, 2), (0, 3), (0, 4), (0, 5), (0, 6), (0, 7), (1, 0), (1, 1), (1, 2), (1, 3), (1, 4), (1, 5), (1, 6), (1, 7), (2, 0), (2, 1), (2, 2), (2, 3), (2, 4), (2, 5), (2, 6), (2, 7), (3, 0), (3, 1), (3, 2), (3, 3), (3, 4), (3, 5), (3, 6), (3, 7), (4, 0), (4, 1), (4, 2), (4, 3), (4, 4), (4, 5), (4, 6), (4, 7), (5, 0), (5, 1), (5, 2), (5, 3), (5, 4), (5, 5), (5, 6), (5, 7), (6, 0), (6, 1), (6, 2), (6, 3), (6, 4), (6, 5), (6, 6), (6, 7), (7, 0), (7, 1), (7, 2), (7, 3), (7, 4), (7, 5), (7, 6), (7, 7)}
    x_index_0 : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
    

In [ ]:
!apt-get install -y -qq coinor-cbc

In [ ]:
opt_cbc = SolverFactory('cbc')

In [ ]:
result = opt_cbc.solve(model_lab6_ex1)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [ ]:
print('\nObjective = ', model_lab6_ex1.objective())

print('\nDecision Variables')
for i in row_indices:
  for j in col_indices:
    if model_lab6_ex1.x[i,j].value != 0:
      print(' number of truck-loads of EVs transported from ',data_csvfile['Warehouses'][i],' to ', data_csvfile.columns[j+1],'=', model_lab6_ex1.x[i,j].value)

print('\nConstraints')
model_lab6_ex1.constraints.display()


Objective =  2192870.0

Decision Variables
 number of truck-loads of EVs transported from  Ahmedabad   to  Jodhpur = 300.0
 number of truck-loads of EVs transported from  Ahmedabad   to  Dibrugarh = 45.0
 number of truck-loads of EVs transported from  Ahmedabad   to  Raipur = 5.0
 number of truck-loads of EVs transported from  Bengaluru  to  Trichy = 120.0
 number of truck-loads of EVs transported from  Bengaluru  to  Vijayawada = 65.0
 number of truck-loads of EVs transported from  Chennai  to  Trichy = 305.0
 number of truck-loads of EVs transported from  Delhi  to  Simla = 95.0
 number of truck-loads of EVs transported from  Delhi  to  Dibrugarh = 85.0
 number of truck-loads of EVs transported from  Kolkata  to  Dibrugarh = 95.0
 number of truck-loads of EVs transported from  Kolkata  to  Agartala = 110.0
 number of truck-loads of EVs transported from  Lucknow  to  Simla = 175.0
 number of truck-loads of EVs transported from  Mumbai  to  Hassan = 205.0
 number of truck-loads of EVs

$ \huge(solution \ \ of \ \ 1.7) $

total cost = 2192870.0

number of truck-loads of EVs transported from  Ahmedabad   to  Jodhpur = 300.0
 
 number of truck-loads of EVs transported from  Ahmedabad   to  Dibrugarh = 45.0
 
 number of truck-loads of EVs transported from  Ahmedabad   to  Raipur = 5.0
 
 number of truck-loads of EVs transported from  Bengaluru  to  Trichy = 120.0
 
 number of truck-loads of EVs transported from  Bengaluru  to  Vijayawada = 65.0
 
 number of truck-loads of EVs transported from  Chennai  to  Trichy = 305.0
 
 number of truck-loads of EVs transported from  Delhi  to  Simla = 95.0
 
 number of truck-loads of EVs transported from  Delhi  to  Dibrugarh = 85.0
 
 number of truck-loads of EVs transported from  Kolkata  to  Dibrugarh = 95.0
 
 number of truck-loads of EVs transported from  Kolkata  to  Agartala = 110.0
 
 number of truck-loads of EVs transported from  Lucknow  to  Simla = 175.0
 
 number of truck-loads of EVs transported from  Mumbai  to  Hassan = 205.0
 
 number of truck-loads of EVs transported from  Mumbai  to  Raipur = 120.0
 
 number of truck-loads of EVs transported from  Hyderabad  to  Vijayawada = 310.0
 
 number of truck-loads of EVs transported from  Hyderabad  to  Raipur = 60.0





$\huge(solution \ \ of \ \ 1.8)$

 if Lucknow-Simla link is disrupted then i change my data i.e .csv file by taking a sufficiently large cost for Lucknow-Simla route since cost is high and we minimize cost so, our pyomo model didn't recomended this route

In [ ]:
data_csvfile.iloc[5,4]= 5000000

In [ ]:
model_lab6_ex1.objective=Objective(expr=sum(data_csvfile.iloc[i,j+1]*model_lab6_ex1.x[i,j] for i in row_indices for j in col_indices),sense=minimize)

This is usually indicative of a modelling error.
To avoid this warning, use block.del_component() and block.add_component().


In [ ]:
result = opt_cbc.solve(model_lab6_ex1)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [ ]:
print('\nObjective = ', model_lab6_ex1.objective())

print('\nDecision Variables')
for i in row_indices:
  for j in col_indices:
    if model_lab6_ex1.x[i,j].value != 0:
      print(' number of truck-loads of EVs transported from ',data_csvfile['Warehouses'][i],'to', data_csvfile.columns[j+1],'=', model_lab6_ex1.x[i,j].value)

print('\nConstraints')
model_lab6_ex1.constraints.display()


Objective =  2380570.0

Decision Variables
 number of truck-loads of EVs transported from  Ahmedabad  to Jodhpur = 300.0
 number of truck-loads of EVs transported from  Ahmedabad  to Raipur = 50.0
 number of truck-loads of EVs transported from  Bengaluru to Trichy = 120.0
 number of truck-loads of EVs transported from  Bengaluru to Vijayawada = 65.0
 number of truck-loads of EVs transported from  Chennai to Trichy = 305.0
 number of truck-loads of EVs transported from  Delhi to Simla = 180.0
 number of truck-loads of EVs transported from  Kolkata to Dibrugarh = 95.0
 number of truck-loads of EVs transported from  Kolkata to Agartala = 110.0
 number of truck-loads of EVs transported from  Lucknow to Dibrugarh = 130.0
 number of truck-loads of EVs transported from  Lucknow to Raipur = 45.0
 number of truck-loads of EVs transported from  Mumbai to Hassan = 205.0
 number of truck-loads of EVs transported from  Mumbai to Simla = 90.0
 number of truck-loads of EVs transported from  Mumbai t

$\huge{solution \ \ of  \ \ 1.9}$

cost  =  2380570.0


 number of truck-loads of EVs transported from  Ahmedabad  to Jodhpur = 300.0

 number of truck-loads of EVs transported from  Ahmedabad  to Raipur = 50.0

 number of truck-loads of EVs transported from  Bengaluru to Trichy = 120.0

 number of truck-loads of EVs transported from  Bengaluru to Vijayawada = 65.0

 number of truck-loads of EVs transported from  Chennai to Trichy = 305.0

 number of truck-loads of EVs transported from  Delhi to Simla = 180.0

 number of truck-loads of EVs transported from  Kolkata to Dibrugarh = 95.0

 number of truck-loads of EVs transported from  Kolkata to Agartala = 110.0

 number of truck-loads of EVs transported from  Lucknow to Dibrugarh = 130.0

 number of truck-loads of EVs transported from  Lucknow to Raipur = 45.0

 number of truck-loads of EVs transported from  Mumbai to Hassan = 205.0

 number of truck-loads of EVs transported from  Mumbai to Simla = 90.0

 number of truck-loads of EVs transported from  Mumbai to Raipur = 30.0

 number of truck-loads of EVs transported from  Hyderabad to Vijayawada = 310.0

 number of truck-loads of EVs transported from  Hyderabad to Raipur = 60.0


 since some route is disturbed so we have to supply that markets by another warehouses that naturally increase the cost of transport 

 in our problem without disrupted route we supply 175 units from lucknow to simla since route is disturbed we have to supply from some other warehouse that increase cost naturally

$\huge{solution \ \ of \ \ 1.10}$

similar to 1.8 here also we increase the cost of disturbed route sufficiently large in .csv file  

since cost is high and we minimize cost so, our pyomo model didn't recomended this route

In [ ]:
data_csvfile.iloc[4,6]=5000000

In [ ]:
data_csvfile.iloc[3,7]=500000

In [ ]:
model_lab6_ex1.objective=Objective(expr=sum(data_csvfile.iloc[i,j+1]*model_lab6_ex1.x[i,j] for i in row_indices for j in col_indices),sense=minimize)

This is usually indicative of a modelling error.
To avoid this warning, use block.del_component() and block.add_component().


In [ ]:
result = opt_cbc.solve(model_lab6_ex1)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [ ]:
print('\nObjective = ', model_lab6_ex1.objective())

print('\nDecision Variables')
for i in row_indices:
  for j in col_indices:
    if model_lab6_ex1.x[i,j].value != 0:
      print(' number of truck-loads of EVs transported from',data_csvfile['Warehouses'][i],' to ', data_csvfile.columns[j+1],'=', model_lab6_ex1.x[i,j].value)

print('\nConstraints')
model_lab6_ex1.constraints.display()


Objective =  2518200.0

Decision Variables
 number of truck-loads of EVs transported from Ahmedabad   to  Jodhpur = 300.0
 number of truck-loads of EVs transported from Ahmedabad   to  Dibrugarh = 50.0
 number of truck-loads of EVs transported from Bengaluru  to  Trichy = 120.0
 number of truck-loads of EVs transported from Bengaluru  to  Vijayawada = 65.0
 number of truck-loads of EVs transported from Chennai  to  Trichy = 305.0
 number of truck-loads of EVs transported from Delhi  to  Simla = 180.0
 number of truck-loads of EVs transported from Kolkata  to  Raipur = 95.0
 number of truck-loads of EVs transported from Kolkata  to  Agartala = 110.0
 number of truck-loads of EVs transported from Lucknow  to  Dibrugarh = 175.0
 number of truck-loads of EVs transported from Mumbai  to  Hassan = 205.0
 number of truck-loads of EVs transported from Mumbai  to  Simla = 90.0
 number of truck-loads of EVs transported from Mumbai  to  Raipur = 30.0
 number of truck-loads of EVs transported fro

$\huge{solution \ \ of \ \ 1.11 }$

cost = 2518200.0

 number of truck-loads of EVs transported from Ahmedabad   to  Jodhpur = 300.0

 number of truck-loads of EVs transported from Ahmedabad   to  Dibrugarh = 50.0

 number of truck-loads of EVs transported from Bengaluru  to  Trichy = 120.0

 number of truck-loads of EVs transported from Bengaluru  to  Vijayawada = 65.0

 number of truck-loads of EVs transported from Chennai  to  Trichy = 305.0

 number of truck-loads of EVs transported from Delhi  to  Simla = 180.0

 number of truck-loads of EVs transported from Kolkata  to  Raipur = 95.0

 number of truck-loads of EVs transported from Kolkata  to  Agartala = 110.0

 number of truck-loads of EVs transported from Lucknow  to  Dibrugarh = 175.0

 number of truck-loads of EVs transported from Mumbai  to  Hassan = 205.0

 number of truck-loads of EVs transported from Mumbai  to  Simla = 90.0

 number of truck-loads of EVs transported from Mumbai  to  Raipur = 30.0

 number of truck-loads of EVs transported from Hyderabad  to  Vijayawada = 310.0

 number of truck-loads of EVs transported from Hyderabad  to  Raipur = 60.0

 $\huge{observation}$
 

 since some route is disturbed so we have to supply that markets by another warehouses that naturally increase the cost of transport 

 in our problem without disrupted route our cost of supply is 2192870.0 after one route (lucknow-simla) distubed the cost is 2380570.0  and now after 3 route is disturbed the cost increase and finally cost becomes 2518200.0
 
   since route is disturbed we have to supply from some other warehouse that increase cost naturally